apps/backend/app/api/routes/data_processing.py

```
@@ -77,22 +77,12 @@ async def process_uploaded_file(
-            import re
-            match = re.search(r'/([^/]+)$', s3_url)
-            if match:
-                file_key = match.group(1)
-            else:
-                raise ValueError(f"Could not extract file key from S3 URL: {s3_url}")
-        
+        # Download file from S3 (parse_s3_url handles all persisted URL shapes)
+        try:
+            _, file_key = parse_s3_url(user_data.s3_url)
+        except ValueError:
+            raise ValueError(f"Could not extract file key from S3 URL: {user_data.s3_url}")
```

Ruff (0.15.15)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)